csharp/src/SeedLang.Ast/Executor.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace SeedLang.Ast {
+  // A executor class to execute an program represented by an AST tree.
+  //
+  // The executor traverses through the AST tree by implementing the interfaces of the AstVisitor
+  // class. The visit method of an expression will return the result of the expression. The return
+  // value of a statement visit method is not used.
+  public sealed class Executor : AstVisitor<BaseValue> {
+    // The global environment of the executor.
+    private readonly Dictionary<string, BaseValue> _globals = new Dictionary<string, BaseValue>();
```

Comment:
Here the field "_globals" looks pretty like a global symbol table. It belongs to the Executor per this definition. This raises a good question:

Since there will be an AST executor as well as a byte code "executor" (or, the interpreter), will both the two executors share the common global symbol tables?

For example, the built-in (native or non-native) functions will be registered to the system. But do we need to register a built-in function twice, to the AST executor and to the byte code executor?

"print" will be a good case for this discussion. If the print function is implemented as a native function, do we need to register it twice?

I guess there can be a singleton LibraryManager (or named in other ways) class to hide all the complexities related to the library functions, including:

- Built-in functions implemented with SeedLang itself (e.g., implemented by SeedPython then compiled into an AST and a byte code)
- Built-in functions implemented with the native environment and expose their interfaces to SeedLan (e.g., implemented in C#)
- Safe third-party functions implemented with SeedLang itself.
- Unsafe third-party function implemented with the native environment, native DLLs, or even the native machine code.

Your thoughts?

Reply:
1. The AST executor and VM executor can not share the common global symbol table, because they use the different value types because of the performance consideration. But the build-in functions they use might be the same ones.
2. The design of following senaria might be:
- Build-in functions implemented with SeedLang itself: they shall be AST trees in the AST executor, and will be compiled to bytecode and stored in a function value of VM.
- Build-in functions implemented with the native enviorment: I considered to create a sandbox class, and store them in the sandbox object. Both AST executor and VM executor will call into the sandbox object.
- Third-party functions implemented with SeedLang itself: shall be the same as the SeedLang build-in functions.
- Unsafe third-party functions implemented with the native enviorment: shall be the same as build-in native functions.

In current phase, I decided to remove the global symbol table, and expose the execution result through a visualizer.

Reply:
Agreed, and just one more clarification: the concept "sandbox" covers two things:

1. As your described, the "sandbox" component runs the built-in and third-party unsafe native extensions and makes sure that the extensions do not violate the safe boundary.
2. Event if a third-party extension is implemented in SeedLang itself, it still requires a sandbox layer (probably a much thinner layer) to protect - for example, attackers may implement an infinite loop to hang the hosting service. The thin sandbox layer should be able to detect such behaviors and solve them.